ujt/transformers.py

```diff
@@ -65,6 +100,116 @@ def apply_views(elements, tag_map, view_list):
+                node_name_sli_map[ujt_id],
+                start_time,
+                end_time,
+            )
+            element_copy["classes"] += f" {change_over_time_class}"
+            output_elements.append(element_copy)
+
+    return output_elements
+
+
+def get_change_over_time_class_from_sli_list(
+    slis: List[SLI], start_time: dt.datetime, end_time: dt.datetime
+) -> str:
+    """Returns the correct coloring/gradient based on the SLI values over time.
+
+    We denote parts of a full styling class as a subclass.
+    These include STATUS_HEALTHY, IMPROVED, etc.
+    We join the subclasses together with underscore as a delimiter to create a full class name,
+    that has an assoiated style.
+
+    Args:
+        slis: A non-empty list of timestamped SLIs.
+
+    Returns:
+        a class name to be added to the element
+    """
+
+    # This shouldn't occur, since the caller uses a defaultdict
+    if slis == []:
+        raise ValueError
+
+    mid_time = start_time + (end_time - start_time) / 2
+    slis_before, slis_after = [], []
+    for sli in slis:
+        if sli.timestamp.ToDatetime() < mid_time:
+            slis_before.append(sli)
+        else:
+            slis_after.append(sli)
+
+    # Compute the average value before and after min_time, and note the appropriate class
+    composite_slis: List[Optional[SLI]] = [None, None]
+    # subclasses holds the class for styling of the left and right half of the element
+    subclasses: List[str] = [constants.NO_DATA_SUBCLASS, constants.NO_DATA_SUBCLASS]
+    for idx, sli_sublist in enumerate([slis_before, slis_after]):
+        if sli_sublist != []:
+            composite_sli = SLI()
+            # copy the slo bound/target values from the input,
+            # these should be constant across all slis in the input list
+            composite_sli.CopyFrom(sli_sublist[0])
+            # compute the average value, this can be more sophisticated
+            composite_sli.sli_value = sum([sli.sli_value for sli in sli_sublist]) / len(
+                sli_sublist
+            )
+            compute_status.compute_sli_status(composite_sli)
+
+            composite_slis[idx] = composite_sli
+            subclasses[idx] = Status.Name(composite_sli.status)
+
+    # Notice that composite_slis can never be [None, None], since we ensured slis != [] above.
+    # Thus, the subclasses can never be equal unless there are two valid composite SLIs
+    if subclasses[0] == subclasses[1]:
+        # Extra asserts here to make mypy happy
+        assert composite_slis[0] is not None
+        assert composite_slis[1] is not None
+
+        slo_target = composite_slis[0].slo_target
+        if abs(slo_target - composite_slis[0].sli_value) < abs(
+            slo_target - composite_slis[1].sli_value
+        ):
+            subclasses[1] += f"_{constants.IMPROVED_SUBCLASS}"
+        else:
+            subclasses[1] += f"_{constants.WORSENED_SUBCLASS}"
+
+    return "_".join(subclasses)
```

Comment:
Would like input on this bit -- this code seems inelegant but not sure how I can improve it.
The "subclass" term doesn't seem very clear but I can't think of a better name.
 
I made a NO_DATA_SUBCLASS for the case where the server only returns data before or after the midpoint time.

I create "composite" SLI objects from the first half and the latter half in order to reuse the compute_sli_status method. (And also it bundles the slo bound/target values nicely).

Finally, I check if the SLI stayed within the same status category and if so, check which direction it moved.